### DISCUSSION TITLE
Script for easier install on mac apple silicon / m1 / m2 / m3

### DISCUSSION THREAD

**Discussion by jonathanstanley** [Comfy-Org/ComfyUI#3283] [Show and tell]
Background: coreML tools are needed to have reasonable speed on a mac. but, the install process for comfy + manager + coreml tools is quite finicky.

After fussing with it for quite some time, I just went ahead and scripted it so the entire process is one command. (if you save the code below as comfy.sh, then the command would be `zsh comfy.sh` from the folder in terminal

```sh
#!/bin/sh

# install homebrew https://brew.sh/
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"

# install python v3.11
# 3.10 is the minimum for coreML tools - https://github.com/aszc-dev/ComfyUI-CoreMLSuite/issues/37#issuecomment-2042594570
# 3.12 drops imp which is required by coreML tools - https://stackoverflow.com/a/77281165/4481226
brew install python@3.11

# install git
brew install git

# copy comfyui files
git clone https://github.com/comfyanonymous/ComfyUI comfyui

# goto comfy folder
cd comfyui

# venv per https://packaging.python.org/en/latest/guides/installing-using-pip-and-virtual-environments/
python3.11 -m venv .venv
source .venv/bin/activate

# install pytorch https://developer.apple.com/metal/pytorch/
pip3.11 install --pre torch torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/nightly/cpu

# install dependencies
pip3.11 install -r requirements.txt

# install comfyui manager
(
    cd custom_nodes
    git clone https://github.com/ltdrdata/ComfyUI-Manager.git
    cd ComfyUI-Manager
    pip3.11 install -r requirements.txt
)

# install coreml tools https://github.com/aszc-dev/ComfyUI-CoreMLSuite
(
    cd custom_nodes
    git clone https://github.com/aszc-dev/ComfyUI-CoreMLSuite.git
    cd ComfyUI-CoreMLSuite
    pip3.11 install -r requirements.txt
)

#start
python3.11 main.py

```

**Comment by TjitseDeVries** [comment]
Tried to run your steps but I get the following error. 

  File "/Users/user/comfyui/main.py", line 134, in <module>
    import comfy.utils
  File "/Users/user/comfyui/comfy/utils.py", line 40, in <module>
    from numpy.dtypes import Float64DType
ModuleNotFoundError: No module named 'numpy.dtypes'

Could you tell which numpy version you are using for this setup?

**Reply by RadonX** [reply]
I recently installed ComfyUI and used Python3.12. Below are my commands. They replace the original thread's instructions between "# venv" and "# install dependencies". I use `uv` to manage Python virtual environment. 

```
# Install uv
brew install uv

# Where you git clone the project
cd ComfyUI

# Create a virtual environment using uv
uv venv

# Install PyTorch (nightly CPU version) per https://github.com/comfyanonymous/ComfyUI?tab=readme-ov-file#apple-mac-silicon
uv pip install --pre torch torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/nightly/cpu

# Install additional project dependencies
uv pip install -r requirements.txt

# Run the main ComfyUI app.
uv run main.py
```

Note that when you install custom_nodes, you also need to install its project dependencies with uv.

```
cd ComfyUI/custom_nodes
git clone https://github.com/ltdrdata/ComfyUI-Manager.git
cd ComfyUI-Manager; uv pip install -r requirements.txt
```

If anything goes wrong, make sure your Python version is 3.12. See how to configure Python version with uv at https://docs.astral.sh/uv/concepts/python-versions/

**Reply by rsl** [reply]
but that drops imp which OP says is needed for ML stuff

**Reply by Black0S** [reply]
🖥️ I built a script that makes installing, launching & updating ComfyUI (and its dependencies) dead simple on **Apple Silicon**.

Feel free to try it out & share your feedback!
👉 https://github.com/Black0S/ComfyUI-Mac-Silicon-Launcher 🙌

**Comment by shubham805** [comment]
Need help in running comfyui on Mac m4.

Getting the error with python3.11

File "/Users/user/comfyui/main.py", line 134, in
import comfy.utils
File "/Users/user/comfyui/comfy/utils.py", line 40, in
from numpy.dtypes import Float64DType
ModuleNotFoundError: No module named 'numpy.dtypes'

with python3.12 - ComfyUI-CoreMLSuite installation error with pip

  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [33 lines of output]
      Traceback (most recent call last):
        File "/Users/jshubh/ComfyUI/.venv/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 389, in <module>
          main()
        File "/Users/jshubh/ComfyUI/.venv/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 373, in main
          json_out["return_val"] = hook(**hook_input["kwargs"])
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/Users/jshubh/ComfyUI/.venv/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 137, in get_requires_for_build_wheel
          backend = _build_backend()
                    ^^^^^^^^^^^^^^^^
        File "/Users/jshubh/ComfyUI/.venv/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 70, in _build_backend
          obj = import_module(mod_path)
                ^^^^^^^^^^^^^^^^^^^^^^^
        File "/opt/homebrew/Cellar/python@3.12/3.12.9/Frameworks/Python.framework/Versions/3.12/lib/python3.12/importlib/__init__.py", line 90, in import_module
          return _bootstrap._gcd_import(name[level:], package, level)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
        File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
        File "<frozen importlib._bootstrap>", line 1310, in _find_and_load_unlocked
        File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
        File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
        File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
        File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
        File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
        File "<frozen importlib._bootstrap_external>", line 999, in exec_module
        File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
        File "/private/var/folders/f1/bj744s8n6j920b7hmr0_65g401bkmf/T/pip-build-env-rkmv_u7d/overlay/lib/python3.12/site-packages/setuptools/__init__.py", line 16, in <module>
          import setuptools.version
        File "/private/var/folders/f1/bj744s8n6j920b7hmr0_65g401bkmf/T/pip-build-env-rkmv_u7d/overlay/lib/python3.12/site-packages/setuptools/version.py", line 1, in <module>
          import pkg_resources
        File "/private/var/folders/f1/bj744s8n6j920b7hmr0_65g401bkmf/T/pip-build-env-rkmv_u7d/overlay/lib/python3.12/site-packages/pkg_resources/__init__.py", line 2172, in <module>
          register_finder(pkgutil.ImpImporter, find_on_path)
                          ^^^^^^^^^^^^^^^^^^^
      AttributeError: module 'pkgutil' has no attribute 'ImpImporter'. Did you mean: 'zipimporter'?
      [end of output]

**Reply by Black0S** [reply]
> Need help in running comfyui on Mac m4.
> 
> Getting the error with python3.11
> 
> File "/Users/user/comfyui/main.py", line 134, in import comfy.utils File "/Users/user/comfyui/comfy/utils.py", line 40, in from numpy.dtypes import Float64DType ModuleNotFoundError: No module named 'numpy.dtypes'
> 
> with python3.12 - ComfyUI-CoreMLSuite installation error with pip
> 
> × Getting requirements to build wheel did not run successfully. │ exit code: 1 ╰─> [33 lines of output] Traceback (most recent call last): File "/Users/jshubh/ComfyUI/.venv/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 389, in main() File "/Users/jshubh/ComfyUI/.venv/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 373, in main json_out["return_val"] = hook(**hook_input["kwargs"]) ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ File "/Users/jshubh/ComfyUI/.venv/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 137, in get_requires_for_build_wheel backend = _build_backend() ^^^^^^^^^^^^^^^^ File "/Users/jshubh/ComfyUI/.venv/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 70, in _build_backend obj = import_module(mod_path) ^^^^^^^^^^^^^^^^^^^^^^^ File "/opt/homebrew/Cellar/python@3.12/3.12.9/Frameworks/Python.framework/Versions/3.12/lib/python3.12/importlib/**init**.py", line 90, in import_module return _bootstrap._gcd_import(name[level:], package, level) ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ File "", line 1387, in _gcd_import File "", line 1360, in _find_and_load File "", line 1310, in _find_and_load_unlocked File "", line 488, in _call_with_frames_removed File "", line 1387, in _gcd_import File "", line 1360, in _find_and_load File "", line 1331, in _find_and_load_unlocked File "", line 935, in _load_unlocked File "", line 999, in exec_module File "", line 488, in _call_with_frames_removed File "/private/var/folders/f1/bj744s8n6j920b7hmr0_65g401bkmf/T/pip-build-env-rkmv_u7d/overlay/lib/python3.12/site-packages/setuptools/**init**.py", line 16, in import setuptools.version File "/private/var/folders/f1/bj744s8n6j920b7hmr0_65g401bkmf/T/pip-build-env-rkmv_u7d/overlay/lib/python3.12/site-packages/setuptools/version.py", line 1, in import pkg_resources File "/private/var/folders/f1/bj744s8n6j920b7hmr0_65g401bkmf/T/pip-build-env-rkmv_u7d/overlay/lib/python3.12/site-packages/pkg_resources/**init**.py", line 2172, in register_finder(pkgutil.ImpImporter, find_on_path) ^^^^^^^^^^^^^^^^^^^ AttributeError: module 'pkgutil' has no attribute 'ImpImporter'. Did you mean: 'zipimporter'? [end of output]

🖥️ I built a script that makes installing, launching & updating ComfyUI (and its dependencies) dead simple on **Apple Silicon**.

Feel free to try it out & share your feedback!
👉 https://github.com/Black0S/ComfyUI-Mac-Silicon-Launcher 🙌

**Comment by Phatcat** [comment]
I was experiencing similar issues to [shubham805](https://github.com/shubham805) and [TjitseDeVries](https://github.com/TjitseDeVries) - Good news guys, the solution was right here in this repo all along.

Quite a few important additions have been added to the 2.8.0 nightly build in regards to apple silicon and it would be lovely to see this on maybe the dev branch of comfyui so more apple folk get to take advantage of workflows like this: https://old.reddit.com/r/comfyui/comments/17q3mov/three_3_sampling_steps_is_all_it_took_for_these/ 
But unfortunately on version 2.6.0 of pytorch it throws this error (along with a bunch of others) on apple silicon: aten::_linalg_solve_ex.result (as tracked here https://github.com/pytorch/pytorch/issues/141287)

For the solution?
As described here https://github.com/comfyanonymous/ComfyUI/issues/6479 ; it's a numpy version issue, so if you upgrade numpy to 1.26.4 (not 2.x) then it should press on. At least it did so for me, I was having the same issue, installing and reinstalling comfyui and python and pytorch over and over again, but now I am happily running pytorch 2.8.0 nightly, even on python 3.10, and that's nested inside Stability Matrix as well even.

Now if I could only get some sort of use out of the Core ML tool's nodes that'd be nice but it seems like that's broken.

**Reply by Black0S** [reply]
> I was experiencing similar issues to [shubham805](https://github.com/shubham805) and [TjitseDeVries](https://github.com/TjitseDeVries) - Good news guys, the solution was right here in this repo all along.
> 
> Quite a few important additions have been added to the 2.8.0 nightly build in regards to apple silicon and it would be lovely to see this on maybe the dev branch of comfyui so more apple folk get to take advantage of workflows like this: https://old.reddit.com/r/comfyui/comments/17q3mov/three_3_sampling_steps_is_all_it_took_for_these/ But unfortunately on version 2.6.0 of pytorch it throws this error (along with a bunch of others) on apple silicon: aten::_linalg_solve_ex.result (as tracked here [pytorch/pytorch#141287](https://github.com/pytorch/pytorch/issues/141287))
> 
> For the solution? As described here #6479 ; it's a numpy version issue, so if you upgrade numpy to 1.26.4 (not 2.x) then it should press on. At least it did so for me, I was having the same issue, installing and reinstalling comfyui and python and pytorch over and over again, but now I am happily running pytorch 2.8.0 nightly, even on python 3.10, and that's nested inside Stability Matrix as well even.
> 
> Now if I could only get some sort of use out of the Core ML tool's nodes that'd be nice but it seems like that's broken.

🖥️ I built a script that makes installing, launching & updating ComfyUI (and its dependencies) dead simple on **Apple Silicon**.

Feel free to try it out & share your feedback!
👉 https://github.com/Black0S/ComfyUI-Mac-Silicon-Launcher 🙌

**Comment by Hussain2906** [comment]
hello i am stuck on adding graph to the model and also i am not able to run the latest version on my mac m1 i have cloned the latest repo but still it shows the old version of 0.3.41 instead of 0.3.45 and also i am doing this with the help of gpt as i dont know how to run this. please help me run this

**Reply by Black0S** [reply]
> hello i am stuck on adding graph to the model and also i am not able to run the latest version on my mac m1 i have cloned the latest repo but still it shows the old version of 0.3.41 instead of 0.3.45 and also i am doing this with the help of gpt as i dont know how to run this. please help me run this

🖥️ I built a script that makes installing, launching & updating ComfyUI (and its dependencies) dead simple on **Apple Silicon**.

Feel free to try it out & share your feedback!
👉 https://github.com/Black0S/ComfyUI-Mac-Silicon-Launcher 🙌
